### PR TITLE
Fixes frame indexing in create_vmd_surface

### LINKED ISSUE #3476: Suggestion of the bugfix hole.py create_vmd_surface function
The function create_vmd_surface is not working if you compute hole analysis starting from a specific frame.

hole.py Line 799:   ` for i, sphpdb in zip(self.frames, self.sphpdbs[self.frames])`
leads to index mismatch between two arrays if self.frames[0] != 0.

The fix that worked for me `for i, sphpdb in zip(self.frames, self.sphpdbs[np.indices(self.frames.shape, sparse = True)])`

Not sure if this is the most elegant solution, but it seems to work :)

### PR DESCRIPTION
**Pull request by IAlibay**
Fixes #3476 

So this _fixes_ the indexing issue, but unfortunately I a) don't have proper access to VMD right now, and b) my knowledge of VMD scripting is more than rusty. - (for @lilyminium or another @MDAnalysis/coredevs  who might know better on how this all works / is able to check) is the way `set triangles` is invoked right now allowing this to work on a non-contiguous trajectory in VMD?

Changes made in this Pull Request:
 - changes the loop over self.frames to index self.results.sphpdbs properly
 - swaps internally the use of self.outfiles and self.sphpdbs instead of self.results.outfiles and self.results.sphpdbs

PR Checklist
------------
 - [x] Tests?
 - [x] Docs?
 - [x] CHANGELOG updated?
 - [x] Issue raised/referenced?

### CODE REVIEW

**Comment by orbeckst** [comment]
I am trying out your patch with VMD 1.9.4a48:

```python
import MDAnalysis as mda
from MDAnalysis.tests import datafiles
from MDAnalysis.analysis import hole2

u = mda.Universe(datafiles.MULTIPDB_HOLE)
BINDIR = "/Users/oliver/hole2/exe"
h2 = hole2.HoleAnalysis(u, executable=BINDIR + "/hole", sos_triangle=BINDIR + "/sos_triangle", sph_process=BINDIR + "/sph_process")
h2.run(start=1, stop=9, step=2)
h2.create_vmd_surface(filename="hole_1_7_2.vmd")
```
This creates hole profiles for frames 1, 3, 5, 7
```python
>>> h2.results.outfiles
array(['hole001.out', 'hole003.out', 'hole005.out', 'hole007.out'],
      dtype=object)
```

Create subsampled trajectory and initial frame for visualization
```python
u.atoms.write("frame0.pdb")
u.atoms.write("frames_1_7_2.trr", frames=list(range(1, 9, 2)))
```

Load in VMD, delete frame 0 (to get a trajectory with the original frames 1, 3, 5, 7).

**ERROR**: VMD cannot load the sliced surface trajectory
```tcl
(tmp) 40 % source hole_1_7_2.vmd
can't read "vmd_frame(-1)": no such element in array
```
The problem is that in VMD trajectories always start at 0 and are indexed with consecutive integer frames.

I'll see if I can change the code.

File: package/MDAnalysis/analysis/hole2/hole.py
Comment (orbeckst) [inline package/MDAnalysis/analysis/hole2/hole.py]: Need to index with `i` for VMD, see https://github.com/MDAnalysis/mdanalysis/pull/3489#issuecomment-1003202993.

**Review by orbeckst — CHANGES_REQUESTED** [review]
Currently fails in VMD, see https://github.com/MDAnalysis/mdanalysis/pull/3489#issuecomment-1003202993.

I'll see if I can quickly fix and test it.

File: package/MDAnalysis/analysis/hole2/hole.py
Comment (IAlibay) [reply package/MDAnalysis/analysis/hole2/hole.py]: Does it make sense to just do that, output N "consecutive frames" (which may have been picked non-consecutively) and let users deal with the mismatch?

Comment (orbeckst) [reply package/MDAnalysis/analysis/hole2/hole.py]: The only way that VMD will "play" the surface movie is by writing consecutive frames. It's the user's responsibility to write the appropriate trajectory as in my example in the comment.

**Comment by orbeckst** [comment]
Currently two other hole tests fail...
```
FAILED MDAnalysisTests/analysis/test_hole2.py::TestHole::test_correct_input - assert False
FAILED MDAnalysisTests/analysis/test_hole2.py::TestHole::test_input_options - assert False
```
Not sure what's happening... looking into it. Sigh.

**Comment by orbeckst** [comment]
crap... my emacs python mode deletes trailing whitespace at the end of the line, which killed tests...

**Comment by orbeckst** [comment]
I can't be bothered to dig into configuring my emacs right now — according to https://stackoverflow.com/questions/14164292/my-emacs-deletes-trailing-white-space-how-can-i-disable-this-behaviour and https://emacs.stackexchange.com/questions/21868/disable-removal-of-trailing-whitespace-in-emacs and https://www.emacswiki.org/emacs/DeletingWhitespace this is too much effort.

Workaround solution:
```
vim test_hole2.py
```
:-p

**Review by orbeckst — APPROVED** [review]

**Comment by orbeckst** [comment]
@IAlibay i approved the PR. Please have a look yourself and recruit another reviewer if necessary. I leave it to you to merge when you’re happy.

**Review by IAlibay — COMMENTED** [review]
Thanks @orbeckst lgtm - I'm going to hand off the final merge decision here to @lilyminium (mostly because I don't fully understand what's going on with those plot3d skip-if deleted tests [it's actually an issue I raised quite some time ago! #3043 ]).

**Comment by lilyminium** [comment]
Thanks @IAlibay, I'll go over this tomorrow -- USA -> AUS was harder than the other way around.

**Comment by lilyminium** [comment]
> The problem is that in VMD trajectories always start at 0 and are indexed with consecutive integer frames.

There are a couple alternative or additional solutions:

* write empty triangles for the frames not in the analysis. That's fairly easy to do, as we have access to the underlying `_trajectory` and therefore know the length
* write out an additional `hole_surface.pdb` trajectory of the analysed frames, to go with the `hole_surface.vmd` of analysed frames currently written in this PR. (However, the written atoms will have to be all the atoms in the universe, as the analysed AtomGroup is an UpdatingAtomGroup. The default selection is "protein" so not sure if this is necessary either)

What do you think? I'm somewhat inclined to think that people will prefer the latter and it's more usable, but I don't really use HOLE much myself. @mavstrikova, do you have any opinions on this?

**Comment by lilyminium** [comment]
Holding off on official approval and merge until further discussion of the question raised in the previous comment, but this looks good to me -- thank you @orbeckst and @IAlibay! I did find an unrelated bug in #3493 .

**Comment by orbeckst** [comment]
I would keep it as simple as possible. I would not automatically write out a trajectory, especially not in the terrible multi-PDB format. I added the docs so that users could do trajectory writing if they wanted to. 

I would also not write a trajectory with triangles at the skipped step as this increases the file for anyone who wants to write their own trajectory. 

To accommodate everyone we would need many additional switches to tune the output behavior. That needs to be tested…

Instead of overengineering and guessing what users would want I’d rather keep it simple and make sure it works. 

In summary: I’d merge as is.

**Comment by IAlibay** [comment]
So I _think_ the main use case we should cover here is:

Creating a HOLE vmd surface with start=10, stop=20, step=5, and then load the full trajectory in VMD whilst also defining the same start/stop/step inputs on file selection (in the molecule file browser) and then also load the newly created surface.

I don't have the ability to test this, but my understanding from @orbeckst's comments is that the way it's currently implemented in this PR is the most likely way to work for this?

I would also agree with @orbeckst that whilst @lilyminium's options would work, they add a certain extra level of complexity on top (plus in the case of writing empty frames kinda breaks the idea of MDA's start/stop/step?).

**Comment by orbeckst** [comment]
> Creating a HOLE vmd surface with start=10, stop=20, step=5, and then load the fully trajectory in VMD but also defined the same start/stop/step inputs on file selection (in the molecule file browser) and then also loaded the newly created surface.

This works if one remembers that VMD's *Last* is **inclusive** so for the following
```python
>>> u.trajectory
<PDBReader ~/mdanalysis/testsuite/MDAnalysisTests/data/1grm_elNemo_mode7.pdb.bz2 with 11 frames of 264 atoms>
>>> with hole2.HoleAnalysis(u, executable=BINDIR+"/hole", sos_triangle=BINDIR+"/sos_triangle", sph_process=BINDIR+"/sph_process") as h2:
        h2.run(start=1, stop=9, step=2)
        h2.create_vmd_surface(filename="hole_192.vmd")
>>> w = mda.Universe("t_192.trr")
>>> w.trajectory
<TRRReader t_192.trr with 4 frames of 264 atoms>
>>> [ts.frame for ts in w.trajectory]
[0, 1, 2, 3]
```
you need to load the original trajectory with 
* First: 1
* **Last: 7** (and not 9!)
* Stride: 2

VMD_load_with_stride
 
Then you can load the surface file
```tcl
vmd > source hole_192.vmd
```
just fine.

**Comment by orbeckst** [comment]
I am adding something to the docs.

**Comment by orbeckst** [comment]
I updated the docs with a force-push.

**Comment by orbeckst** [comment]
Not sure what the codcov error is about
```
[2022-01-03T17:55:52.383Z] ['error'] There was an error running the uploader: Error uploading to https://codecov.io: Error: There was an error fetching the storage URL during POST: 404 - Not Found - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
```

**Comment by IAlibay** [comment]
> Not sure what the codcov error is about
> 
> ```
> [2022-01-03T17:55:52.383Z] ['error'] There was an error running the uploader: Error uploading to https://codecov.io: Error: There was an error fetching the storage URL during POST: 404 - Not Found - {'detail': ErrorDetail(string='Unable to locate build via Github Actions API. Please upload with the Codecov repository upload token to resolve issue.', code='not_found')}
> ```

Looks like there's something going on with GH Actions - some weird timeout errors on random runners. I've restarted CI, let's see what happens.

**Comment by lilyminium** [comment]
Failing test is due to using `assert_equal` to compare floats in testing mdamath.angle; this should probably be `assert_allclose`.

**Comment by orbeckst** [comment]
Alright, I'll procrastinate a little bit and fix it... thanks for identifying the issue @lilyminium

**Comment by IAlibay** [comment]
> Alright, I'll procrastinate a little bit and fix it... thanks for identifying the issue @lilyminium

Already got a PR open for this @orbeckst (#3495)

**Review by lilyminium — APPROVED** [review]

**Pull request merged**
